feat: Upgrade Weights & Biases callback

parambharat opened the pull request:
# What does this PR do?
This PR adds a few new functionalities to the Weights & Biases Callback

- Logs Peft and Lora Config to wandb if present
- Adds model parameter counts to wandb config and artifact metadata
- Adds on_predict methods to log prediction metrics
- Prints the model architecture to a file alongside the wandb artifact
- Logs initial and final models to the wandb artifact to full reproducibility
- Adds steps and epoch aliases to checkpoint artifacts

- Here's a [link](https://wandb.ai/parambharat/test-transformers/artifacts/model/model-rg4pcjcv/v3) to the what the logged artifacts look like
- Here's a run [overview page](https://wandb.ai/parambharat/test-transformers/runs/rg4pcjcv/overview?workspace=user-parambharat) with added config and metadata for the run with peft configs logged


## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?

## Who can review?

- trainer: @muellerzr and @pacman100

parambharat committed:
feat: add peft config to wandb if it exists in the model

parambharat committed:
feat: add peft config to wandb if it exists in the model

parambharat committed:
feat: add model parameter count to wandb config and model metadata

parambharat committed:
feat: add model parameter count to wandb config and model metadata

parambharat committed:
feat: add metrics on prediction to wandb

parambharat committed:
feat: add metrics on prediction to wandb

parambharat committed:
feat: add model architecture to the model artifact

parambharat committed:
feat: add model architecture to the model artifact

parambharat committed:
feat: add initial model and architecture to the model artifact on setup

parambharat committed:
feat: add initial model and architecture to the model artifact on setup

parambharat committed:
Merge branch 'main' into wandb/callback-upgrade

parambharat committed:
feat: add markdown badge to model card

parambharat committed:
feat: add parameters for peft models and model card badge

parambharat committed:
feat: add parameters for peft models and model card badge

parambharat committed:
Merge branch 'main' into wandb/callback-upgrade

parambharat committed:
refactor: change checkpoints to log and model and rename initial to base

parambharat committed:
refactor: change checkpoints to log and model and rename initial to base

parambharat committed:
feat: add step and epoch aliases to the checkpoints

parambharat committed:
feat: add step and epoch aliases to the checkpoints

parambharat committed:
chore: run fixup and style fixes

parambharat committed:
chore: run fixup and style fixes

parambharat committed:
Merge branch 'main' into wandb/callback-upgrade

muellerzr reviewed (approved):
Thanks! Overall looks good to me

ArthurZucker commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -785,20 +841,35 @@ def on_train_end(self, args, state, control, model=None, tokenizer=None, **kwarg
                     if (args.run_name is None or args.run_name == args.output_dir)
                     else f"model-{self._wandb.run.name}"
                 )
+                # add the model architecture to a separate text file
```
I am not 100% sure I understand why the architecture would change over time during training?

parambharat replied:
> I am not 100% sure I understand why the architecture would change over time during training.

@ArthurZucker : Here, we log the model architecture to a text file in the model artifact. Some of our users have requested this. This helps to audit and reproduce the experiment easily.

ArthurZucker commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -816,18 +887,30 @@ def on_save(self, args, state, control, **kwargs):
                 for k, v in dict(self._wandb.summary).items()
                 if isinstance(v, numbers.Number) and not k.startswith("_")
             }
+            checkpoint_metadata["model/num_parameters"] = self._wandb.config["model/num_parameters"]
 
             ckpt_dir = f"checkpoint-{state.global_step}"
             artifact_path = os.path.join(args.output_dir, ckpt_dir)
             logger.info(f"Logging checkpoint artifacts in {ckpt_dir}. ...")
             checkpoint_name = (
-                f"checkpoint-{self._wandb.run.id}"
+                f"model-{self._wandb.run.id}"
                 if (args.run_name is None or args.run_name == args.output_dir)
-                else f"checkpoint-{self._wandb.run.name}"
+                else f"model-{self._wandb.run.name}"
```
is that not breaking? could we keep model and `aliases=["model"]` ?

parambharat replied:
> is that not breaking? could we keep model and `aliases=["model"]` ?

@ArthurZucker : This doesn't break existing functionality. Currently, we create separate Weights & Biases Artifacts for models and checkpoints. However, this results in multiple artifacts for the same experiment. This instead creates a single artifact with model and checkpoint aliases, allowing the users to log models and checkpoints to the same artifact.

ArthurZucker reviewed (commented):
Thanks have a few small questions but good otherwise!

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -816,18 +887,30 @@ def on_save(self, args, state, control, **kwargs):
+                artifact, aliases=["checkpoint", f"epoch_{round(state.epoch, 2)}", f"global_step_{state.global_step}"]
+            )
+
+    def on_predict(self, args, state, control, metrics, **kwargs):
+        if self._wandb is None:
+            return
+        if not self._initialized:
+            self.setup(args, state, **kwargs)
+        if state.is_world_process_zero:
+            metrics = rewrite_logs(metrics)
+            self._wandb.log(metrics)
```
> could you explain why we need this?

@ArthurZucker : This allows users to log model predictions and metrics to W&B dashboard when they use `model.predict` . This is quite useful in evaluation runs.

parambharat commented:
@muellerzr and @ArthurZucker : Thank you for the review and comments. I have responded to the review comments with some explanations. Let me know if it looks good.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29125). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

parambharat committed:
Merge branch 'main' into wandb/callback-upgrade

# Conflicts:
#	src/transformers/integrations/integration_utils.py

parambharat commented:
@muellerzr : Resolved :+1:

muellerzr commented:
cc @amyeroberts for a final review as she's on watch this week 🤗

amyeroberts commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -756,6 +761,57 @@ def setup(self, args, state, model, **kwargs):
+                    elif isinstance(model, (torch.nn.Module, PushToHubMixin)) and hasattr(model, "base_model"):
+                        print(model, file=f)
+
+                for f in Path(temp_dir).glob("*"):
```
This seems like it could be really slow - taking all the files in the current directory and saving them out as artefacts

parambharat replied:
This should not be an issue. This is how current artifact logging works. While the files are added to the artifact the logging happens async using the wandb service. This allows the experiment to continue while the artifact files are uploaded to W&B by the wandb service. 

Also, having all the files in a model/checkpoint allows users to make reproducible runs.

amyeroberts reviewed (commented):
Thanks for working on this! 

Just a few small-ish questions and comments

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -817,18 +888,30 @@ def on_save(self, args, state, control, **kwargs):
             logger.info(f"Logging checkpoint artifacts in {ckpt_dir}. ...")
             checkpoint_name = (
-                f"checkpoint-{self._wandb.run.id}"
+                f"model-{self._wandb.run.id}"
```
While we name the artifact as model, for checkpoints we add an additional alias `checkpoint`. This ensures that an artifact for a single run includes both model and checkpoints. Here's an example with different versions

![image](https://github.com/huggingface/transformers/assets/12809212/d6422695-7c5c-42d0-8203-9620b4c2a6a4)

parambharat committed:
fix: address review comments related to DRY and naming consistency

parambharat committed:
fix: address review comments related to DRY and naming consistency

parambharat committed:
Merge branch 'main' of github.com:parambharat/transformers into wandb/callback-upgrade

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -817,18 +888,30 @@ def on_save(self, args, state, control, **kwargs):
             artifact.add_dir(artifact_path)
-            self._wandb.log_artifact(artifact, aliases=[f"checkpoint-{state.global_step}"])
+            self._wandb.log_artifact(
+                artifact, aliases=["checkpoint", f"epoch_{round(state.epoch, 2)}", f"global_step_{state.global_step}"]
```
Fixed. Using underscores consistently across names and aliases now.

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -786,20 +842,35 @@ def on_train_end(self, args, state, control, model=None, tokenizer=None, **kwarg
+
+                        model.summary(print_fn=print_to_file)
+                    elif isinstance(model, (torch.nn.Module, PushToHubMixin)) and hasattr(model, "base_model"):
+                        print(model, file=f)
```
Currently, we intend to support logging architectures for `PreTrainedModel`, `TFPreTrainedModel`, and `PEFT` models. The current logic handles this. We can add Flax and other model types based on usage and user feature requests.

parambharat replied:
Good catch, - Handled this and changed it to use `config.get` instead.

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -756,6 +761,57 @@ def setup(self, args, state, model, **kwargs):
+                    type="model",
+                    metadata={
+                        "model_config": model.config.to_dict() if hasattr(model, "config") else None,
+                        "num_parameters": model.num_parameters(),
```
Good catch, - Handled this and changed it to use config.get instead.

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -756,6 +761,57 @@ def setup(self, args, state, model, **kwargs):
+
+                        model.summary(print_fn=print_to_file)
+                    elif isinstance(model, (torch.nn.Module, PushToHubMixin)) and hasattr(model, "base_model"):
+                        print(model, file=f)
```
Same as above, 

Currently, we intend to support logging architectures for `PreTrainedModel`, `TFPreTrainedModel`,  and `PEFT` models. The current logic handles this. We can add Flax and other model types based on usage and user feature requests.

parambharat commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -756,6 +761,57 @@ def setup(self, args, state, model, **kwargs):
+                )
+                model.save_pretrained(temp_dir)
+                # add the architecture to a separate text file
+                with open(f"{temp_dir}/model_architecture.txt", "w+") as f:
```
Thanks.  Have abstracted it out to a separate function `save_model_architecture_to_file`

parambharat commented:
@amyeroberts . Thanks for the detailed comments and review. I have addressed all the comments and responded to them. Please take a look and let me known if there is anything pending.

parambharat commented:
Hi @amyeroberts , Just checking in on this PR. Let me know if you have any comments and if my responses to the review comments are accepted and resolve them.

amyeroberts commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -829,18 +896,30 @@ def on_save(self, args, state, control, **kwargs):
             artifact.add_dir(artifact_path)
-            self._wandb.log_artifact(artifact, aliases=[f"checkpoint-{state.global_step}"])
+            self._wandb.log_artifact(
+                artifact, aliases=[f"epoch_{round(state.epoch, 2)}", f"checkpoint_global_step_{state.global_step}"]
```
Still inconsistent here with `-` vs `_`

amyeroberts reviewed (approved):
Thanks for iterating on this! 

I'm still not 100% sure about some aspects e.g. saving out the model arch at the end of training, as there's been demonstrated successful runs with outputs and integrations are maintained by contributors rather than the transformers team, I'm happy for this to be merge, with any other maintenance / follow up handle by the W&B team

amyeroberts commented:
@parambharat Can you rebase on main to include any upstream changes?

parambharat committed:
Merge branch 'main' of github.com:parambharat/transformers into wandb/callback-upgrade

parambharat commented:
> @parambharat Can you rebase on main to include any upstream changes?

merged main into the branch and pushed the changes.

stevhliu committed:
[docs] Big model loading (#29920)

* update

* feedback

ArthurZucker committed:
[`generate`] fix breaking change for patch (#29976)

* fix bug and add tests

* nit

* otherway to get the cur len instead of attention mask

* more places where this might have been broken

* nit

* oups

* inputs_embeds vs input_embeds

* test generated outptus

* style

* nit

* fix

* skip failing biogpt

hovnatan committed:
Fix 29807 sinusoidal positional encodings in Flaubert, Informer and XLM (#29904)

* Fix sinusoidal_embeddings in FlaubertModel

* Fix for Informer

* Fix for XLM

* Move sinusoidal emb for XLM

* Move sinusoidal emb for Flaubert

* Small cleanup

* Add comments on tests code copied from

* Add with Distilbert->

SunMarc committed:
[bnb] Fix bug in `_replace_with_bnb_linear` (#29958)

fix bug

giganttheo committed:
Adding FlaxNoRepeatNGramLogitsProcessor (#29677)

* fix issue with logit processor in beam search in Flax

* adding FlaxNoRepeatNGramLogitsProcessor class + unit test

* style correction and code verification

* add FlaxNoRepeatNGramLogitsProcessor to the test_processor_list and test_processor_list_jitted tests

* fix an issue where ngrams are banned only if they appear ==1 time + update description of get_previous_ngrams

* replace non-jit compatible masking of ngrams that are not yet generated with jittable version

* Revert "fix issue with logit processor in beam search in Flax"

This reverts commit 09b70d7e4dc32d0cc4db61af09a835a9cd238b50.

* add FlaxNoRepeatNGramLogitsProcessor to _get_logits_processor

* change the method of casting to boolean of banned tokens indices

* fix code style

* remove some useless operations + significantly faster computation of update indices using jax.lax.fori_loop

* remove useless loop iterations

* set some variables that were calculated and used multiple times

* fix format

ylacombe committed:
Add Flash Attention 2 support to Musicgen and Musicgen Melody (#29939)

* add FA2 to o.g Musicgen

* make style

* add FA2 support to Musicgen Melody

* add generation FA2 tests to o.g Musicgen

* make style and fix copies

* add Musicgen to FA2 docs + deprecate list

* add sdpa supports to Musicgen's

* make style and fix copies

* refactor attention implementation arguments

* add Copied from to sdpa tests

* add copied form in sdpa tests melody

* add copied for FA2 generation tests

* add FA2 inference copied from

* make style

windsonsea committed:
[Docs] Make an ordered list prettier in add_tensorflow_model.md (#29949)

msublee committed:
Fix `skip_special_tokens` for `Wav2Vec2CTCTokenizer._decode` (#29311)

* Fix skip_special_tokens process for Wav2Vec2CTCTokenizer._decode

* Fix skip_special_tokens for Wav2Vec2CTCTokenizer._decode

* Exclude pad_token filtering since it is used as CTC-blank token

* Add small test for skip_special_tokens

* Update decoding test for added new token

Narsil committed:
Hard error when ignoring tensors. (#27484) (#29906)

* Hard error when ignoring tensors. (#27484)

* [WIP] Hard error when ignoring tensors.

* Better selection/error when saving a checkpoint.

- Find all names we should normally drop (those are in the transformers
  config)
- Find all disjoint tensors (for those we can safely trigger a copy to
  get rid of the sharing before saving)
- Clone those disjoint tensors getting rid of the issue
- Find all identical names (those should be declared in the config
  but we try to find them all anyway.)
- For all identical names:
  - If they are in the config, just ignore them everything is fine
  - If they are not, warn about them.
- For all remainder tensors which are shared yet neither identical NOR
  disjoint. raise a hard error.

* Adding a failing test on `main` that passes here.

* We don't need to keep the subfolder logic in this test.

* Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Add small tests.

* Dead variable.

* Fixup.

* Fixing tied_Weights_keys on generic models.

* Fixup + T5 encoder/decoder tying (with different layers)

* Code quality.

* Dynamic member.

* trigger

* Fixing encoder name for other types of encoder/decoder combos.

* Fix scoping.

* Update .github/workflows/self-scheduled.yml

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Fixing the tied_weights after the call.

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>
Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

gante committed:
Generate: fix logits processors doctests (#29718)

* fix norm

* fix logits processors doctests

mariosasko committed:
Fix `remove_columns` in `text-classification` example (#29351)

ydshieh committed:
Update `tests/utils/tiny_model_summary.json` (#29941)

update

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

fxmarty committed:
Make EncodecModel.decode ONNX exportable (#29913)

* fix encodec onnx export for musicgen

* simplification

* fix quality

* better style

Commit:
Fix Swinv2ForImageClassification NaN output (#29981)

To address the issue of NaN logit outputs for certain combinations
of the `image_size`, `patch_size` and `depths` configuration
parameters, an assertion was made to ensure that the resulting
`window_size` field in the model's Self Attention class is greater
than 1, preventing divisions by zero in the normalization of
`relative_coords_table`.

Fix: #28675

jklj077 committed:
Fix Qwen2Tokenizer (#29929)

qwen2: fixed tokens starting with # in slow tokenizer; add tests

Co-authored-by: jklj077 <17811943+jklj077@users.noreply.github.com>

cifkao committed:
Fix `kwargs` handling in `generate_with_fallback` (#29225)

* Fix generate_with_fallback **kwargs

* Change pop to get

* Delete keys from kwargs to prevent overriding generation_config

* Revert to passing kwargs by reference, but make a (shallow) copy

* dict -> copy.copy

* Add test_whisper_longform_multi_batch_beam

cifkao committed:
Fix probability computation in `WhisperNoSpeechDetection` when recomputing scores (#29248)

* Fix is_scores_logprobs in WhisperNoSpeechDetection

* Add test_whisper_longform_no_speech_detection

* Fix typo

zucchini-nlp committed:
Fix vipllava for generation (#29874)

* fix vipllava generation

* consistent llava code

* revert llava tests changes

stevhliu committed:
[docs] Fix audio file (#30006)

new audio file

zucchini-nlp committed:
Superpoint imports fix (#29898)

quick fix

ArthurZucker committed:
[`Main CIs`] Fix the red cis  (#30022)

* fix

* sort imports

muellerzr committed:
Make clearer about zero_init requirements (#29879)

* Docstring to note about zero init

* Check for accelerate

* Change conditional return

* Tweak

* Add new accelerate-specific zero3 check

* Fix import

* Revert to RTFM

* Update src/transformers/modeling_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jla524 committed:
Enable multi-device for efficientnet (#29989)

feat: enable mult-idevice for efficientnet

byi8220 committed:
Add a converter from mamba_ssm -> huggingface mamba (#29705)

* implement convert_mamba_ssm_checkpoint_to_pytorch

* Add test test_model_from_mamba_ssm_conversion

* moved convert_ssm_config_to_hf_config to inside mamba_ssm_available check

* fix skipif clause

* moved skips to inside test since skipif decorator isn't working for some reason

* Added validation

* removed test

* fixup

* only compare logits

* remove weight rename

* Update src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* nits

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

byi8220 committed:
[`ProcessingIdefics`] Attention mask bug with padding (#29449)

* Defaulted IdeficsProcessor padding to 'longest', removed manual padding

* make fixup

* Defaulted processor call to padding=False

* Add padding to processor call in IdeficsModelIntegrationTest as well

* Defaulted IdeficsProcessor padding to 'longest', removed manual padding

* make fixup

* Defaulted processor call to padding=False

* Add padding to processor call in IdeficsModelIntegrationTest as well

* redefaulted padding=longest again

* fixup/doc

saurabhdash2512 committed:
Refactor Cohere Model (#30027)

* changes

* addressing comments

* smol fix

amyeroberts commented:
@parambharat Could you try again rebasing? We can't merge until all the tests are 🟢  At the moment, these don't seem related to this PR, but I haven't observed them failing on other open PRs

ydshieh committed:
Add `whisper` to `IMPORTANT_MODELS` (#30046)

Add whisper

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
skip `test_encode_decode_fast_slow_all_tokens` for now (#30044)

skip test_encode_decode_fast_slow_all_tokens for now

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sywangyi committed:
if output is tuple like facebook/hf-seamless-m4t-medium, waveform is … (#29722)

* if output is tuple like facebook/hf-seamless-m4t-medium, waveform is the first element

Signed-off-by: Wang, Yi <yi.a.wang@intel.com>

* add test and fix batch issue

Signed-off-by: Wang, Yi <yi.a.wang@intel.com>

* add dict output support for seamless_m4t

Signed-off-by: Wang, Yi <yi.a.wang@intel.com>

---------

Signed-off-by: Wang, Yi <yi.a.wang@intel.com>

AdamLouly committed:
Fix mixtral ONNX Exporter Issue. (#29858)

* fix mixtral onnx export

* fix qwen model

NielsRogge committed:
[Trainer] Allow passing image processor (#29896)

* Add image processor to trainer

* Replace tokenizer=image_processor everywhere

parambharat committed:
chore: update and rebase with upstream main

# Conflicts:
#	src/transformers/integrations/integration_utils.py

parambharat committed:
chore: update and rebase with upstream main

# Conflicts:
#	src/transformers/integrations/integration_utils.py

parambharat committed:
chore: update and rebase with upstream main

# Conflicts:
#	src/transformers/integrations/integration_utils.py

parambharat committed:
chore: run make fixup

parambharat commented:
@amyeroberts : It looks like rebasing with main pulled a bunch of commits and changes into the branch. This was unintentional. I'm closing this PR and raising a new one with just the changes relevant to the integration.

parambharat commented:
@amyeroberts and @muellerzr : Please find the new PR [here](https://github.com/huggingface/transformers/pull/30135)

amyeroberts commented:
@parambharat For future PRs, when many commits are added like this, it normally indicates that the force push wasn't used when pushing the rebased branch to the remote. As rebasing is effectively rewriting the history, it's necessary to do `git push -f` to have the rebase changes reflected.
